lib/rspec/core/example_group.rb

```diff
@@ -340,6 +340,16 @@ def self.include_examples(name, *args, &block)
         find_and_eval_shared("examples", name, caller.first, *args, &block)
       end
 
+      # Adds an example to the example group
+      def self.add_example example
+        examples << example
+      end
+
+      # Removes an example from the example group
+      def self.remove_example example
+        examples.delete example
+      end
```

Comment:
There are several methods that contain memoized derived state from the set of examples, such as https://github.com/rspec/rspec-core/blob/b8f0270f57253dbf02087b120824d84af6e35dce/lib/rspec/core/example_group.rb#L410-L413 and https://github.com/rspec/rspec-core/blob/b8f0270f57253dbf02087b120824d84af6e35dce/lib/rspec/core/example_group.rb#L583-L587.

For those to work properly, the memoized value will need to be cleared when an example is added or removed.  That's a big reason why the `examples` attribute isn't public -- if users directly mutate the array we can't keep those things in sync.